gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdkManager.java

```diff
@@ -171,4 +182,40 @@ private void addCaCert(Project project, Path javaHome, String alias, String caCe
                     alias, javaHome, output.toString(StandardCharsets.UTF_8)));
         }
     }
+
+    /**
+     * Abstraction around locking access to a file or directory, by creating another file with a
+     * matching name, and '.lock' extension. Note that the underlying file locking mechanism is
+     * left to the filesystem, so we must be careful to work within its bounds. For example,
+     * POSIX file locks apply to a process, so within the process we must ensure synchronization
+     * separately.
+     */
+    private static final class PathLock implements Closeable {
+        private static final Striped<Lock> JVM_LOCKS = Striped.lock(16);
+        private final Closer closer;
+
+        PathLock(Path path) throws IOException {
+            this.closer = Closer.create();
+            try {
+                Lock jvmLock = JVM_LOCKS.get(path);
+                jvmLock.lock();
+                closer.register(jvmLock::unlock);
+                Files.createDirectories(path.getParent());
+                FileChannel channel = closer.register(FileChannel.open(
+                        path.getParent().resolve(path.getFileName() + ".lock"),
+                        StandardOpenOption.CREATE,
+                        StandardOpenOption.WRITE));
+                FileLock fileLock = channel.lock();
```

Comment:
Fancy - didn't know you could get process-level file locks like this. What does it do on windows/NTFS, out of curiosity?

Reply:
On windows, the in-process (non-fs lock) is actually unnecessary since unlike unix, locks aren't owned by entire processes! No reason to remove that lock, though, since it would only increase the code complexity. Ideally we could get windows tests running in #69, however there are some hurdles with pjf+checkstyle